lib/SPIRV/libSPIRV/SPIRVModule.cpp

```diff
@@ -1069,15 +1078,69 @@ SPIRVValue *SPIRVModuleImpl::addNullConstant(SPIRVType *Ty) {
 
 SPIRVValue *SPIRVModuleImpl::addCompositeConstant(
     SPIRVType *Ty, const std::vector<SPIRVValue *> &Elements) {
+  SPIRVConstantComposite *Res;
+  const size_t MaxNumElements = 65535 - 3;
+  const size_t NumElements = Elements.size();
+  if (NumElements > MaxNumElements &&
+      isAllowedToUseExtension(ExtensionID::SPV_INTEL_long_constant_composite)) {
+    for (uint64_t J = 0; J < NumElements; J += MaxNumElements) {
+      auto Start = Elements.begin() + J;
+      auto End = ((J + MaxNumElements) > NumElements) ? Elements.end()
+                                                      : Start + MaxNumElements;
+      std::vector<SPIRVValue *> Slice(Start, End);
+      if (J == 0) {
+        Res = static_cast<SPIRVConstantComposite *>(
+            addCompositeConstant(Ty, Slice));
+      } else {
+        SPIRVConstantComposite::ContinuedInstType Continued =
+            static_cast<SPIRVConstantComposite::ContinuedInstType>(
+                addCompositeConstantContinuedINTEL(Slice));
+        Res->addContinuedInstruction(Continued);
+      }
+    }
+    return Res;
+  }
   return addConstant(new SPIRVConstantComposite(this, Ty, getId(), Elements));
```

Comment:
Just a suggestion for rearrangement of code (there might be errors because I didn't compile and test it). I think it is simpler to read. You may ignore this suggestion if don't think so.
```suggestion
  const size_t MaxNumElements = 65535 - 3;
  const size_t NumElements = Elements.size();
  if (NumElements <= MaxNumElements ||
      !isAllowedToUseExtension(ExtensionID::SPV_INTEL_long_constant_composite)) {
    return addConstant(new SPIRVConstantComposite(this, Ty, getId(), Elements));
  }

  auto Start = Elements.begin();
  auto End = Start + MaxNumElements;
  std::vector<SPIRVValue *> Slice(Start, End);
  auto *Res = static_cast<SPIRVConstantComposite *>(addCompositeConstant(Ty, Slice));

  do {
    Start = End;
    End = ((Elements.end() - End) > MaxNumElements) ? End + MaxNumElements
                                                    : Elements.end();
    Slice.assign(Start, End);
    SPIRVConstantComposite::ContinuedInstType Continued =
        static_cast<SPIRVConstantComposite::ContinuedInstType>(
            addCompositeConstantContinuedINTEL(Slice));
    Res->addContinuedInstruction(Continued);
  } while ((End - Start) > MaxNumElements);

  return Res;
```

Reply:
I'm not sure that it is simpler. Any other opinions?

Reply:
Suggestion from Mikhail allows to reduce amount of nested code blocks and from my point of view it is readable enough, but I don't have a strong opinion and fine with both ways

Reply:
I already applied another variant of this suggestion.